3rdParty/fuerte/src/H2Connection.cpp

```diff
@@ -745,16 +768,16 @@ void H2Connection<T>::setTimeout() {
         auto& me = static_cast<H2Connection<T>&>(*s);
         // cancel expired requests
         auto now = std::chrono::steady_clock::now();
-        auto it = me._streams.begin();
-        while (it != me._streams.end()) {
-          if (it->second->expires < now) {
-            FUERTE_LOG_DEBUG << "HTTP2-Request timeout\n";
-            it->second->invokeOnError(Error::Timeout);
-            it = me._streams.erase(it);
-          } else {
-            it++;
+        std::vector<int32_t> expired;
+        for (auto const& pair : me._streams) {
+          if (pair.second->expires < now) {
+            expired.push_back(pair.first);
           }
         }
+        std::for_each(expired.begin(), expired.end(), [&](auto sid) {
+          auto strm = me.eraseStream(sid);
+          strm->invokeOnError(Error::Timeout);
+        });
 
```

Comment:
why is it necessary to first copy all streams into a vector, and then iterate over the vector separately?
can't we optimize this to
```
        for (auto it = me._streams.begin(); it != me._streams.end(); /* nada */) {
          if (it->second->expires < now) {
            auto tmp = std::move(it->second);
            it = me._streams.erase(it);
            if (tmp) {
              uint32_t cc = _streamCount.fetch_sub(1, std::memory_order_relaxed);
              FUERTE_ASSERT(cc > 0);
              tmp->invokeOnError(Error::Timeout);
            }
          } else {
            ++it;
          }
        }
        if (me._streams.empty()) {
          startPing();
        }
```
this would be kind of redundant however, so it could use a bit of cleanup. however, it seems unnecessary to create the extra vector and to find the stream in the map once again although we are already pointing at it when iterating over the expired streams.